src/components/tl/cuda/alltoallv/alltoallv_ce.c

```diff
@@ -0,0 +1,402 @@
+            !ucc_tl_cuda_team_topo_is_direct(&team->super, team->topo,
+                                             UCC_TL_TEAM_RANK(team), i)) {
+            continue;
+        }
+        peer_sync = TASK_SYNC(task, i);
+        cache     = ucc_tl_cuda_get_cache(team, i);
+        if (ucc_unlikely(!cache)) {
+            status = UCC_ERR_NO_MESSAGE;
+            goto exit_err;
+        }
+        status = ucc_tl_cuda_map_memhandle(
+            peer_sync->mem_info_src.ptr, peer_sync->mem_info_src.length,
+            peer_sync->mem_info_src.handle,
+            &task->alltoallv_ce.peer_map_addr_src[i], cache);
+        if (UCC_OK != status) {
+            ucc_error("ucc_cuda_ipc_map_memhandle failed");
+            return UCC_ERR_INVALID_PARAM;
+        }
+        status = ucc_tl_cuda_map_memhandle(
+            peer_sync->mem_info_dst.ptr, peer_sync->mem_info_dst.length,
+            peer_sync->mem_info_dst.handle,
+            &task->alltoallv_ce.peer_map_addr_dst[i], cache);
+        if (UCC_OK != status) {
+            ucc_error("ucc_cuda_ipc_map_memhandle failed");
+            return UCC_ERR_INVALID_PARAM;
+        }
+    }
+    return UCC_OK;
+
+exit_err:
+    return status;
+}
+
+ucc_status_t ucc_tl_cuda_alltoallv_ce_post_copies(ucc_tl_cuda_task_t *task)
+{
+    ucc_tl_cuda_team_t         *team = TASK_TEAM(task);
+    ucc_rank_t                  rank = UCC_TL_TEAM_RANK(team);
+    ucc_tl_cuda_sync_t         *sync = TASK_SYNC(task, rank);
+    ucc_tl_cuda_sync_t         *peer_sync;
+    ucc_ee_executor_t          *exec;
+    void                       *src, *dst;
+    ucc_ee_executor_task_t    **exec_task;
+    size_t                      data_size, data_displ;
+    size_t                      rdt_size = ucc_dt_size(task->alltoallv_ce.rdt);
+    size_t                      sdt_size = ucc_dt_size(task->alltoallv_ce.sdt);
+    ucc_rank_t                  i, peer, psrc, pdst;
+    ucc_status_t                status;
+    ucc_ee_executor_task_args_t exec_args;
+
+    task->alltoallv_ce.num_posted = 0;
+    status = ucc_coll_task_get_executor(&task->super, &exec);
+    if (ucc_unlikely(status != UCC_OK)) {
+        goto exit;
+    }
+
+    for (i = 0; i < UCC_TL_TEAM_SIZE(team); i++) {
+        peer = (rank + i) % UCC_TL_TEAM_SIZE(team);
+        if (!ucc_tl_cuda_team_topo_is_direct(&team->super, team->topo, rank,
+                                             peer)) {
+            continue;
+        }
+        peer_sync = TASK_SYNC(task, peer);
+        if (peer == rank) {
+            src = task->alltoallv_ce.sbuf;
+        } else {
+            src = PTR_OFFSET(task->alltoallv_ce.peer_map_addr_src[peer],
+                             peer_sync->mem_info_src.offset);
+            CUDA_CHECK_GOTO(
+                cudaStreamWaitEvent(team->stream,
+                                    sync->data[peer].ipc_event_remote, 0),
+                exit, status);
+        }
+        data_size = sdt_size * task->alltoallv_ce.get_count(
```

Comment:
for alltoall, get_count function returns number of elements in dst buffer which might be different from number of elements in src buffer, so need to multiple by rdt_size

Reply:
hmm...in the case of a remote process uses different datatype, I think both rdt_size and sdt_size could be wrong. I will take your suggestion below to save size and offset in bytes to solve this problem.

Reply:
is it correct? If the src is rank itself (i.e., `peer==rank`), then it should be `sdt_size`. If src is from remote GPU, it should be remote's `sdt_size` which might not be the same as the local `sdt_size`/`rdt_size`, if I understand the semantic correctly. I like the idea you mentioned below of saving number of bytes instead elements to make things easier.

Reply:
right, let's just save number of bytes in shm segment 😊